### PR TITLE
Add filter to prohibit using encoding for columns at foreign tables

### PR DESCRIPTION
**Pull request by red1452**
If set `ENCODING` to `gp_default_storage_options`, this metadata will be
added to foreign tables. Column `ENCODING` is useless for foreign tables,
but it will be printed from metadata if `pg_dump` is executed.

This patch fixes:
1) Commands parser to exclude saving `ENCODING` if table is foreign.
2) `pg_dump`: not to print `ENCODING` for foreign tables even if there is
`ENCODING` at table's metadata.

## Steps to reproduce:

1. change GUC value
   ```
   gpconfig -c 'gp_default_storage_options' -v 'appendonly=true,blocksize=32768,compresstype=zstd,compresslevel=3,checksum=true,orientation=column'
   gpstop -u
   ```
2. create foreign table in the `test` database 
   ```
   CREATE FOREIGN DATA WRAPPER dummy;
   create server test foreign data wrapper dummy;
   create foreign table test (a int) server test;
   ```
3. change GUC values back
   ```
   gpconfig -r 'gp_default_storage_options'
   gpstop -u
   ```
4. create dump with table
   ```
   pg_dump -t public.test test
   ```
## Actual behavior

Output contains invalid foreign definition. This definition can't be used to recreate this table.
```sql
...
CREATE FOREIGN TABLE public.test (
    a integer ENCODING (compresstype=zstd,blocksize=32768,compresslevel=3)
)
SERVER test
;
...
```
## Result after fixes
```sql
...
CREATE FOREIGN TABLE public.test (
    a integer
)
SERVER test
;
...
```

### Another way to get the problem
1) Create demo cluster
2) Set `gp_default_storage_options`
```
gpconfig -c 'gp_default_storage_options' -v 'appendonly=true,blocksize=32768,compresstype=zstd,compresslevel=3,checksum=true,orientation=column'
gpstop -u
```
3) Add foreign table to `regression`
```
createdb regression
psql regression <<EOF
    CREATE FOREIGN DATA WRAPPER dummy;
    CREATE SERVER test FOREIGN DATA WRAPPER dummy;
    CREATE FOREIGN TABLE foreign_table (
        a integer,
        b integer,
        c integer
    ) SERVER test;
    ALTER FOREIGN TABLE foreign_table drop column b;
EOF
```
4) Run tests for `pg_upgrade`:
```
createdb isolation2test
make -C contrib/pg_upgrade installcheck
```
5) Result:
Test fails with error:
```
Restoring database schemas in the new cluster
  regression                                                
*failure*

Consult the last few lines of "pg_upgrade_dump_16384.log" for
the probable cause of the failure.
Failure, exiting

real    0m3.982s
user    0m0.091s
sys     0m0.317s
ERROR: Failure encountered in upgrading qd node
~/gpdb/adb-6.x-gpdb/contrib/pg_upgrade ~/gpdb/adb-6.x-gpdb/contrib/pg_upgrade/tmp_check/upgrade/qd ~/gpdb/adb-6.x-gpdb/contrib/pg_upgrade
make: *** [Makefile:49: installcheck] Error 1
make: Leaving directory '/home/evgeniy/gpdb/adb-6.x-gpdb/contrib/pg_upgrade'
```
Incorrect command, which generated by `pg_dump`:
```
CREATE FOREIGN TABLE ""public"".""foreign_table"" (
    ""a"" integer ENCODING (compresstype=zstd,blocksize=32768,compresslevel=3),
    ""........pg.dropped.2........"" INTEGER /* dummy */ ENCODING (compresstype=zstd,blocksize=32768,compresslevel=3),
    ""c"" integer ENCODING (compresstype=zstd,blocksize=32768,compresslevel=3)
)
SERVER ""test""
;
ALTER FOREIGN TABLE ""public"".""foreign_table"" DROP COLUMN ""........pg.dropped.2........"";
```
Error log:
```
ENCODING clause only supported with column oriented tables
```

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by RekGRpth** [comment]
It turns out that if I dumped the old version and it wrote down the encoding for external tables, then I can’t restore such a dump?

**Comment by RekGRpth** [comment]
> * Place, where table's metadata is created, to exclude save `ENCODING` from global options (`gp_default_storage_options`)
> 
> * `pg_dump`: not to print `ENCODING` for foreign tables even if there is `ENCODING` at table's metadata.

Whether it is possible to make tests and on these two cases?

File: src/backend/commands/tablecmds.c [hunk 1]
```diff
@@ -974,7 +974,8 @@ DefineRelation(CreateStmt *stmt, char relkind, Oid ownerId, char relstorage, boo
 		cooked_constraints = list_concat(cooked_constraints, newCookedDefaults);
 	}
 
-	if (stmt->attr_encodings)
+	/* Encoding clause is useless for foreign tables */
+	if (stmt->attr_encodings && !RelationIsForeign(rel))
```
Comment (Stolb27) [inline src/backend/commands/tablecmds.c]: What about other kinds of relations (heap tables, external tables)? How are they handled?

**Comment by red1452** [comment]
> It turns out that if I dumped the old version and it wrote down the encoding for external tables, then I can’t restore such a dump?

Dump with foreign table with `ENCODING` at column is incorrect. If you try to apply such dump, you will get error:
```
ERROR: ENCODING clause only supported with column oriented tables
```
Because `ENCODING` do not apply to columns at foreign tables.

File: src/backend/commands/tablecmds.c (hunk 1 above)
Comment (red1452) [reply src/backend/commands/tablecmds.c]: There is not any problem with other kind of relations. Extra column options do not apply if table is not `appendonly=true` or not `orientation=column`.
```
test=# create table test3 (a int) with (appendoptimized=false);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE TABLE
test=# \q
...
CREATE TABLE public.test3 (
    a integer
)
WITH (appendonly='false')
 DISTRIBUTED BY (a);
```
```
test=# CREATE EXTERNAL TABLE ext_expenses (name text, date date, 
amount float4, category text, description text) 
LOCATION ( 
'file://seghost1/dbfast/external/expenses1.csv'
)
FORMAT 'CSV' ( HEADER );
NOTICE:  HEADER means that each one of the data files has a header row
CREATE EXTERNAL TABLE
...
CREATE EXTERNAL TABLE public.ext_expenses (
    name text,
    date date,
    amount real,
    category text,
    description text
) LOCATION (
    'file://seghost1/dbfast/external/expenses1.csv'
) ON ALL 
FORMAT 'csv' (delimiter E',' null E'' escape E'"' quote E'"' header)
ENCODING 'UTF8';
```

**Comment by RekGRpth** [comment]
> If you try to apply such dump, you will get error:
> 
> ```
> ERROR: ENCODING clause only supported with column oriented tables
> ```

Then why is it necessary [to issue a new error](https://github.com/arenadata/gpdb/pull/565/files#diff-5bd59ecc8991bacaefd56f7fe986287b8d664e62566eb3588c3845d7625cacf1R4613-R4618), while the error was issued before?

**Comment by red1452** [comment]
> Then why is it necessary [to issue a new error](https://github.com/arenadata/gpdb/pull/565/files#diff-5bd59ecc8991bacaefd56f7fe986287b8d664e62566eb3588c3845d7625cacf1R4613-R4618), while the error was issued before?

The main goal of patch is to fix that `pg_dump` generates valid table definition. I did it by fixed saving metadata while table is created and by added check to `pg_dump` (if table is foreign, not to print column `ENCODING` to command).
Also I found the problem, if `appendonly=true,orientation=column` is set to `gp_default_storage_options`, foreign table with column `ENCODING` will be created success. Because I added new check to commands parser with new error message.

File: src/backend/parser/parse_utilcmd.c [hunk 2]
```diff
@@ -4610,6 +4610,12 @@ transformAttributeEncoding(List *stenc, CreateStmt *stmt, CreateStmtContext *cxt
 		newenc = lappend(newenc, c);
 	}
 
+	/* Do not apply encoding if table is foreign */
+	if (found_enc && IsA(stmt, CreateForeignTableStmt))
+		ereport(ERROR,
```
Comment (InnerLife0) [inline src/backend/parser/parse_utilcmd.c]: It seems we already have a correct error message. Can we just modify `can_enc` assignment to trigger it?
```
bool can_enc = is_aocs(stmt->options) && !IsA(stmt, CreateForeignTableStmt);
```

File: src/test/regress/sql/column_compression.sql [hunk 3]
```diff
@@ -810,3 +810,54 @@ CREATE TABLE co_large_and_bulk_content(a INT,
...
+-- Check that created foreign table does not apply gp_default_storage_options
+-- Expect: success
+create foreign table foreign_table (
+    a integer
+) server test;
+create role foreign_table_owner;
+alter foreign table public.foreign_table owner to foreign_table_owner;
+\d+ foreign_table
+
+-- Check that pg_dump generates valid table defenition
```
Comment (InnerLife0) [inline src/test/regress/sql/column_compression.sql]: _definition_

File: src/backend/commands/tablecmds.c (hunk 1 above)
Comment (InnerLife0) [reply src/backend/commands/tablecmds.c]: `stmt->attr_encodings` is already filled by `transformAttributeEncoding()` function, which you fixed. Why do we need extra check here?

File: src/backend/commands/tablecmds.c [hunk 4]
```diff
@@ -7838,8 +7839,9 @@ ATExecAddColumn(List **wqueue, AlteredTableInfo *tab, Relation rel,
 	/*
 	 * For AO/CO tables, always store an encoding clause. If no encoding
 	 * clause was provided, store the default encoding clause.
+	 * Useless for foreign tables.
 	 */
-	if (RelationIsAoCols(rel))
+	if (RelationIsAoCols(rel) && !RelationIsForeign(rel))
```
Comment (InnerLife0) [inline src/backend/commands/tablecmds.c]: In which cases relation can be both, column-oriented and foreign? How can I trigger this condition manually to check it?

File: src/bin/pg_dump/pg_dump.c [hunk 5]
```diff
@@ -13676,7 +13676,8 @@ dumpTableSchema(Archive *fout, TableInfo *tbinfo)
 						appendPQExpBufferStr(q, " NOT NULL");
 
 					/* Column Storage attributes */
-					if (tbinfo->attencoding[j] != NULL)
+					if (tbinfo->relkind != RELKIND_FOREIGN_TABLE &&
```
Comment (InnerLife0) [inline src/bin/pg_dump/pg_dump.c:13750]: Am I right we do this to correctly backup old foreign tables defined with encoding?
I wonder can we trigger pg_dump.c:13643 also?

**Comment by InnerLife0** [comment]
The general question: Why did you decide to place tests inside `column_compression.sql`, especially including `postgres_fdw` to it?
We already have a fine `foreign_data.sql` with dummy data wrapper.

File: src/backend/parser/parse_utilcmd.c (hunk 2 above)
Comment (red1452) [reply src/backend/parser/parse_utilcmd.c]: I think this is two different error, which should have two different error messages.

File: src/backend/commands/tablecmds.c (hunk 1 above)
Comment (red1452) [reply src/backend/commands/tablecmds.c]: `transformAttributeEncoding()` applies `ENCODING` which is set at table columns, but not filters if it is set at `gp_default_storage_options`.

File: src/bin/pg_dump/pg_dump.c (hunk 5 above)
Comment (InnerLife0) [reply src/bin/pg_dump/pg_dump.c:13750]: I just asked can we somehow trigger pg_dump.c:13643.
What is the test case?

File: src/backend/commands/tablecmds.c (hunk 1 above)
Comment (InnerLife0) [reply src/backend/commands/tablecmds.c]: `transformAttributeEncoding()` covers all cases. Default options are placed inside `stenc` argument.
I did a mistake in comment about `can_enc` assignment, the correct one is:
```
bool can_enc = is_aocs(stmt->options) && !cxt->isforeign;
```

File: src/backend/parser/parse_utilcmd.c (hunk 2 above)
Comment (InnerLife0) [reply src/backend/parser/parse_utilcmd.c]: We don't have any special message for other cases, which can trigger this error. What's special with the foreign tables?

File: src/test/regress/sql/foreign_data.sql [hunk 6]
```diff
@@ -544,7 +544,54 @@ REASSIGN OWNED BY regress_test_role TO regress_test_role2;
...
 
+------------------------------------------------------------------------------
+-- Test to check column encoding at foreign tables.
+-- Not use column encoding at foreign tables.
+------------------------------------------------------------------------------
+
+create database foreign_tables_test_db;
+\c foreign_tables_test_db
+
+create extension postgres_fdw;
```
Comment (InnerLife0) [inline src/test/regress/sql/foreign_data.sql]: Take a look around other adb regress tests. We don't use postgres_fdw not just like that. It's a separate software from contrib with own set of tests.
In my last comment I said we have a dummy data wrapper here.

**Comment by red1452** [comment]
### New commit message:
Add filter to prohibit using encoding for columns at foreign tables (#565)

If set `ENCODING` to `gp_default_storage_options`, this metadata will be
added to foreign tables. Column `ENCODING` is useless for foreign tables,
but it will be printed from metadata if `pg_dump` is executed.

This patch fixes:
1) Commands parser to exclude saving `ENCODING` if table is foreign.
2) `pg_dump`: not to print `ENCODING` for foreign tables even if there is
`ENCODING` at table's metadata.

File: src/test/regress/sql/foreign_data.sql [hunk 7]
```diff
@@ -544,7 +544,54 @@ REASSIGN OWNED BY regress_test_role TO regress_test_role2;
...
+SET gp_default_storage_options = 'appendonly=true,blocksize=32768,compresstype=zstd,compresslevel=3,checksum=true,orientation=column';
+CREATE FOREIGN TABLE foreign_table (
+    a integer ENCODING (compresstype=zstd,blocksize=32768,compresslevel=3)
+) SERVER test;
+
+-- Check that created foreign table does not apply gp_default_storage_options
+-- Expect: success
+CREATE FOREIGN TABLE foreign_table (
+    a integer
+) SERVER test;
```
Comment (bimboterminator1) [inline src/test/regress/sql/foreign_data.sql:573]: We assume, that it's obvious for the user, that `gp_default_storage_options` will be ignored? Writing some kind of `NOTICE` will definitely be redundant?

Comment (InnerLife0) [reply src/test/regress/sql/foreign_data.sql:573]: Defaults are ignored even for heap tables (without any notice as well).

File: src/bin/pg_dump/pg_dump.c (hunk 5 above)
Comment (red1452) [reply src/bin/pg_dump/pg_dump.c:13750]: If do steps from [issue](https://github.com/greenplum-db/gpdb/issues/6193):
```
gpconfig -c 'gp_default_storage_options' -v 'appendonly=true,blocksize=32768,compresstype=zstd,compresslevel=3,checksum=true,orientation=column'
gpstop -u

createdb regression
psql regression <<EOF
    CREATE FOREIGN DATA WRAPPER dummy;
    CREATE SERVER test FOREIGN DATA WRAPPER dummy;
    CREATE FOREIGN TABLE foreign_table (
        a integer,
        b integer,
        c integer
    ) SERVER test;
    ALTER FOREIGN TABLE foreign_table drop column b;
EOF

# Run the upgrade test
createdb isolation2test
make -C contrib/pg_upgrade installcheck
```
At log command will be generated:
```
CREATE FOREIGN TABLE ""public"".""foreign_table"" (
    ""a"" integer ENCODING (compresstype=zstd,blocksize=32768,compresslevel=3),
    ""........pg.dropped.2........"" INTEGER /* dummy */ ENCODING (compresstype=zstd,blocksize=32768,compresslevel=3),
    ""c"" integer ENCODING (compresstype=zstd,blocksize=32768,compresslevel=3)
)
SERVER ""test""
;
```
and test will fail:
```
ENCODING clause only supported with column oriented tables
```

**Comment by red1452** [comment]
I updated the description.

**Review by InnerLife0 — APPROVED** [review]
And here is how table script from last error looked like before latest commit.
```
CREATE FOREIGN TABLE "public"."foreign_table" (
    "a" integer,
    "........pg.dropped.2........" INTEGER /* dummy */ ENCODING (compresstype=zstd,blocksize=32768,compresslevel=3),
    "c" integer
)
SERVER "test"
;
```

**Review by bimboterminator1 — APPROVED** [review]

**Comment by red1452** [comment]
I updated commit message and squashed commits without code changing.

**Pull request merged**